Onboarding: Loading step uses step container v2

zaguiini opened the pull request:
Related to https://github.com/Automattic/wp-calypso/issues/101194.

## Proposed Changes

The checkout loading states are going to be addressed in https://github.com/Automattic/wp-calypso/issues/101201.

The main `document/index.jsx` page uses the default loader, and applies an override so the loading bar is centered. This is because if we render `<Loading />` with the `useStepContainerV2` flag from the backend it doesn't work as it doesn't have the window dimensions yet, so we need to work with what we have for now 😁 

~~A BLOCKER: The post-checkout loading state.~~
Philip: I don't think this is a blocker for this PR. It can be done in an other PR.

## Testing Instructions

Assert that only the loading screens in the onboarding and site setup flow changed.

zaguiini committed:
Use StepContainerV2 in onboarding and site-setup loading screens

zaguiini committed:
Use the loading screen outside of the flow i.e. boot and flow loading pages

zaguiini committed:
Do not query the URL within Document renderer

alshakero commented:
> I'm noticing the TopBar flashes during redirects, and there's also some weird blinking at the top, with a scrollbar (4s).

This is fixed [here](https://github.com/Automattic/wp-calypso/pull/101369) btw. And this is not a scroll bar, it's the loading skeleton.

alshakero commented on client/components/loading/index.tsx:
```diff
@@ -1,15 +1,34 @@
+} ) => {
+	if ( useStepContainerV2 ) {
+		return (
+			<AsyncLoad
+				require="./step-container-v2-loading"
```
I think it's bit of a bad idea to async load a loader. What will show when the loader is suspending?

zaguiini replied:
Makes sense, I'll make it sync then 👍🏻

alshakero reviewed (commented):
Just a drive by comment trying to help.

zaguiini committed:
Synchronously load the component

zaguiini committed:
Hide header and centralize progress bar if loading the stepcontainerv2

autumnfjeld commented:
@zaguiini  Made a video of my test flow.   https://screen.studio/share/I2Bcf8Az

https://github.com/user-attachments/assets/d7b86644-2337-4fab-a0a3-97949ca2ca48


Also noticed the top bar flashing, so we should test with the fix @alshakero  mentions.

I will prioritize testing for all this StepV2 work for  the rest of this week.

alshakero commented:
I merged [my PR](https://github.com/Automattic/wp-calypso/pull/101369) and rebased this for more accurate testing.

zaguiini committed:
Fix unit tests

zaguiini committed:
Only show the title if it exists

zaguiini committed:
Do not load stepcontainerv2 dynamically within domains

zaguiini committed:
Do not load stepcontainerv2 dynamically within plans

zaguiini committed:
Fix layout shift on initial load

paulopmt1 reviewed (commented):
We're having the same amount of blinks from prod now but in different stages. If we can fix those first ones, it would be great!

- Here's how it looks like on local (3 blinks in different stages of the scenarios below):

	https://github.com/user-attachments/assets/80f15976-df6f-41fb-bff5-450312e256d5

- Local with the feature flag disabled (3 blinks):

	https://github.com/user-attachments/assets/ef99fe62-bf75-41a1-9729-cce82a82fcd2

- Prod (3 blinks):

	https://github.com/user-attachments/assets/609cd771-5c03-46ff-b945-88e90da17780

p-jackson commented on client/components/loading/index.tsx:
```diff
@@ -7,12 +8,32 @@ interface LoadingProps {
+	subtitle,
+	progress,
+	className,
+	useStepContainerV2,
```
I know we might be in the final testing stages, so maybe this is something for a follow up. But Adding a boolean prop doesn't feel right here. There's basically nothing shared between the v1 and v2 versions of this loading component.

Better to make it a separate components IMO. You don't end up with weird things like sometimes the `className` and `subtitle` props do something, and sometimes they don't.

p-jackson commented on client/landing/stepper/declarative-flow/internals/steps-repository/processing-step/index.tsx:
```diff
@@ -220,6 +221,15 @@ const ProcessingStep: React.FC< ProcessingStepProps > = function ( props ) {
+		return (
+			<>
+				<DocumentHead title={ __( 'Processing' ) } />
+				<Loading title={ getCurrentMessage() } progress={ progress } useStepContainerV2 />
```
Looks like we've dropped the subtitle and the "show Jetpack Powered logo" options. Are we just relying on the fact that we happen to know all flows for which `shouldUseStepContainerV2` return true there's never a subtitle at the moment?

Maybe that's fine, the only place I could find using a subtitle is this copy-site processing step:
https://github.com/Automattic/wp-calypso/blob/c12d53f98fc0adcfb35f86d0ba2592bd908b1c11/client/landing/stepper/declarative-flow/internals/steps-repository/processing-step-copy-site-flow/index.tsx#L11

autumnfjeld replied:
I'd be wary of removing those unless it was a explicit design decision.

zaguiini replied:
Removing the subtitle was an explicit design decision: W9xI27S6Swvw5Ku21EbZvn-fi-9588_13066#1183813152

p-jackson replied:
And after looking, the `isJetpackPowered` flag is only on the tailored newsletter flow. So I'm not worried about ignoring that case. We'll either retire that flow or redesign it I suspect.

p-jackson commented on client/landing/stepper/utils/get-flow-from-url.ts:
```diff
@@ -1,16 +1,24 @@
-import { matchPath } from 'react-router-dom';
```
I think this is being removed because this module now has to be included on the server.

Importing from `'react-router-dom'` was actually wrong. You can import the `matchPath` from the `'react-router'` package so it can be imported on the server and the client.

The reason I like the idea of continuing to use the route matching logic from React Router is because we want it to match the logic used by the client-side stepper router. And [the stepper internals use `getFlowFromURL` too](https://github.com/Automattic/wp-calypso/blob/c12d53f98fc0adcfb35f86d0ba2592bd908b1c11/client/landing/stepper/index.tsx#L87). So it seems to keep all that logic matching.

zaguiini replied:
Ah great catch, I completely forgot the agnostic package 👍🏻

p-jackson commented on client/landing/stepper/utils/get-flow-from-url.ts:
```diff
@@ -1,16 +1,24 @@
-import { matchPath } from 'react-router-dom';
+export const getFlowFromURL = (
+	url = window.location.pathname,
```
I do agree with adding the ability to pass in an optional pathname though, so that this utility function can be used on the server.

Maybe rename `url` to `pathname` so it's clear this should just be the path fragment?

p-jackson commented on client/signup/steps/domains/index.jsx:
```diff
@@ -1421,8 +1422,7 @@ export class RenderDomainsStep extends Component {
 				return (
-					<AsyncLoad
-						require="./async-domain-step-wrapper"
+					<StepContainerV2DomainsStep
```
I only created `async-domain-step-wrapper` to facilitate the async loading. I think it'd be better to roll this into the main domain step component, since all it does now is prop drilling.

Can be a follow up though.

p-jackson reviewed (commented):
Big lift! Great job @zaguiini 

I'm going to have a bit of a think about preventing that initial flicker when you first navigate from the `/sites` page to the onboarding.

FWIW the LOHP links directly to `/setup/onboarding`. So it's only the existing user experience that flickers.

autumnfjeld commented:
Nice work smoothing some of the experience. 

https://github.com/user-attachments/assets/293dd2a1-c480-43cd-92d7-ed9eebdf0a28


> The blinks in the first video are caused by the redirect from /setup -> /setup/onboarding/domains. I'm not entirely sure there's a way to fix it. Maybe we can default to onboarding as the detected flow if the user entered /setup in the address bar?

> FWIW the LOHP links directly to /setup/onboarding. So it's only the existing user experience that flickers.

The flicker also happens on:
* /setup/site-setup/processing to  /setup/site-setup/goals

zaguiini committed:
Revert to using react-router in getFlowFromURL and rename param to pathname instead of url

zaguiini committed:
Remove unnecessary domains component

zaguiini committed:
Remove unnecessary plans component

zaguiini committed:
Redelegate the decision of showing the vertically centered layout to StepContainerV2

zaguiini committed:
Split the loading into its own component

p-jackson commented:
> I know the blinks are pre-existing, so not mandatory to fix with this work. It would be great to put some brainpower on this for a near future fix. It's one the "janky experiences" that has been mentioned a few times in CfTs and walkthroughs. Related comments on this kind of blinking or flickering have come from both Oscar & Ian--meaning these details are worth thinking about for our continuing onboarding quality work. 🧠 💪

@autumnfjeld, @zaguiini figured out a way to fix the flickering of the WordPress logo while I acted as rubber duck. It requires a rework of the top bar component, so he'll do it in a different PR. I'm going to fix the unit tests and get this PR merged.

p-jackson committed:
Fix unit tests related to getFlowFromURl changes

Reinstates code which defaults to `''` when ther's no
window.location.pathname. Some tests don't pass an explicit pathname to
`getFlowFromUrl` _and_ they also don't have window.location.pathname
defined.

p-jackson commented:
Found an issue where we've inadvertently broken the width of the plans step

**Before this PR**
![CleanShot 2025-03-27 at 11 52 10@2x](https://github.com/user-attachments/assets/ac38818e-177e-4ab2-802a-9d0ab6b43953)

**This PR**
![CleanShot 2025-03-27 at 11 51 54@2x](https://github.com/user-attachments/assets/f59af898-a971-4df3-ab50-3a9268a965bf)

p-jackson reviewed (approved):
Great job, this is looking good and I'm going to merge.

The flicker in the top-right isn't resolved, but Luis is working on that in a follow up PR (involves some server-side rendering changes)
